docs/source/perf_train_gpu_one.mdx

```diff
@@ -0,0 +1,184 @@
+
+    ```python
+    from torch.cuda.amp import autocast
+    with autocast(dtype=torch.bfloat16):
+        loss, outputs = ...
+    ```
+
+
+### Gradient Accumulation
+
+Gradient accumulation allows you to train a much larger batch than can fit into your GPU's memory. It also speeds up the training because it updates the weights less frequently.
+
+The science of gradient accumulation is explained [here](performance#gradient-accumulation).
+
+You will need to experiment to find the best number of steps to accumulate for. In the following examples we will use 4 steps.
+
+Activation:
+
+- HF Trainer-based examples: add `--gradient_accumulation_steps 4` to the command line arguments.
+- Custom HF Trainer-based program, pass one of:
+
+    ```python
+    TrainingArguments(gradient_accumulation_steps=4)
+    ```
+- `accelerate`:  use: ... (XXX: Sylvain/Leandro?)
+
+- Custom training loop: This is somewhat complex but you can study how this is implemented in [HF Trainer](
+https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py) - simply search for `gradient_accumulation_steps` in the code.
+
+
+### Gradient Checkpointing
+
+Gradient checkpointing saves memory while trading off some 20-30% of throughput.
+
+The science of gradient checkpointing is explained [here](performance#gradient-checkpointing).
+
+Activation:
+
+- HF Trainer-based examples: add `--gradient_checkpointing` to the command line arguments.
+- Custom HF Trainer-based program, pass one of:
+
+    ```python
+    TrainingArguments(gradient_checkpointing=True)
+    ```
+- `accelerate`:  use: ... (XXX: Sylvain/Leandro?)
+
+- Custom training loop: To implement see [torch.utils.checkpoint.checkpoint](https://pytorch.org/docs/stable/checkpoint.html).
+
+
+
+### Optimizer
+
+Some optimizers require a lot more memory than others.
+
+The science of optimizer's memory usage and which optimizers to choose when are explained [here](performance#optimizer).
+
+
+Activation:
+
+- HF Trainer-based examples: add `--optim` to the command line arguments, followed by the desired optimizer. e.g. one of `adamw_hf`, `adamw_torch`, `adamw_torch_xla`, `adamw_apex_fused`, `adafactor`
+- Custom HF Trainer-based program, pass one of the optimizers listed above like so:
+
+    ```python
+    TrainingArguments(optim="adam_torch")
+    ```
+- `accelerate`:  use: ... (XXX: Sylvain/Leandro?)
+
+- Custom training loop: This is somewhat complex but you can study how this is implemented in [HF Trainer](
+https://github.com/huggingface/transformers/blob/master/src/transformers/trainer.py) - simply search for `optimizer` in the code.
+
+
+### Deepspeed ZeRO
+
+The in-depth details on how to use Deepspeed can be found [here](main_classes/deepspeed).
+
+First, a quick decision tree:
+
+1. Model fits onto a single GPU and you have enough space to fit a small batch size - you don't need to use Deepspeed as it'll only slow things down in this use case.
```

Comment:
I would add "for training" as you need space for the gradients, optimizer states etc.